Mathlib/Data/Zmod/Defs.lean

```diff
@@ -0,0 +1,188 @@
+-/
+import Mathlib.Algebra.NeZero
+import Mathlib.Data.Nat.ModEq
+import Mathlib.Data.Fintype.Lattice
+
+/-!
+# Definition of `Zmod n` + basic results.
+
+This file provides the basic details of `Zmod n`, including its commutative ring structure.
+
+## Implementation details
+
+This used to be inlined into `Data.Zmod.Basic`. This file imports `CharP.Basic`, which is an
+issue; all `CharP` instances create an `Algebra (Zmod p) R` instance; however, this instance may
+not be definitionally equal to other `Algebra` instances (for example, `GaloisField` also has an
+`Algebra` instance as it is defined as a `SplittingField`). The way to fix this is to use the
+forgetful inheritance pattern, and make `CharP` carry the data of what the `smul` should be (so
+for example, the `smul` on the `GaloisField` `CharP` instance should be equal to the `smul` from
+its `SplittingField` structure); there is only one possible `Zmod p` algebra for any `p`, so this
+is not an issue mathematically. For this to be possible, however, we need `CharP.Basic` to be
+able to import some part of `Zmod`.
+
+-/
+
+
+namespace Fin
+
+/-!
+## Ring structure on `Fin n`
+
+We define a commutative ring structure on `Fin n`, but we do not register it as instance.
+Afterwords, when we define `Zmod n` in terms of `Fin n`, we use these definitions
+to register the ring structure on `Zmod n` as type class instance.
+-/
+
+
+open Nat.ModEq Int
+
+/-- Multiplicative commutative semigroup structure on `fin n`. -/
+instance (n : ℕ) : CommSemigroup (Fin n) :=
+  { inferInstanceAs (Mul (Fin n)) with
+    mul_assoc := fun ⟨a, ha⟩ ⟨b, hb⟩ ⟨c, hc⟩ =>
+      Fin.eq_of_veq
+        (calc
+          a * b % n * c ≡ a * b * c [MOD n] := (Nat.mod_modEq _ _).mul_right _
+          _ ≡ a * (b * c) [MOD n] := by rw [mul_assoc]
+          _ ≡ a * (b * c % n) [MOD n] := (Nat.mod_modEq _ _).symm.mul_left _
+          )
+    mul_comm := Fin.mul_comm }
+
+private theorem left_distrib_aux (n : ℕ) : ∀ a b c : Fin n, a * (b + c) = a * b + a * c :=
+  fun ⟨a, ha⟩ ⟨b, hb⟩ ⟨c, hc⟩ =>
+  Fin.eq_of_veq
+    (calc
+      a * ((b + c) % n) ≡ a * (b + c) [MOD n] := (Nat.mod_modEq _ _).mul_left _
+      _ ≡ a * b + a * c [MOD n] := by rw [mul_add]
+      _ ≡ a * b % n + a * c % n [MOD n] := (Nat.mod_modEq _ _).symm.add (Nat.mod_modEq _ _).symm
+      )
+
+/-- Commutative ring structure on `Fin n`. -/
+instance (n : ℕ) [NeZero n] : CommRing (Fin n) :=
+  { Fin.instAddMonoidWithOneFin n, Fin.addCommGroup n,
+    Fin.instCommSemigroupFin n with
+    one_mul := Fin.one_mul
+    mul_one := Fin.mul_one
+    left_distrib := left_distrib_aux n
+    right_distrib := fun a b c => by
+      rw [mul_comm, left_distrib_aux, mul_comm _ b, mul_comm],
+    -- porting note: new, see
+    -- https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/ring.20vs.20Ring/near/322876462
+    zero_mul := Fin.zero_mul
+    mul_zero := Fin.mul_zero }
+
+end Fin
+
+/-- The integers modulo `n : ℕ`. -/
+def Zmod : ℕ → Type
```

Comment:
```suggestion
def ZMod : ℕ → Type
```
and throughout (including the filename), as requested by @hrmacbeth [here](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Naming.20convention/near/322551595)